lib/ide/ide/src/controller.rs

```diff
@@ -0,0 +1,377 @@
+            pub fn new(handle:utils::handle::WeakHandle<$data_type>) -> $handle_typename {
+                $handle_typename(handle)
+            }
+        }
+
+        impl IsWeakHandle for $handle_typename {
+            type Data = $data_type;
+            fn weak_handle(&self) -> Weak<RefCell<Self::Data>> {
+                self.0.clone()
+            }
+        }
+    };
+}
+
+mod ide {
+    use super::*;
+    use json_rpc::test_util::transport::mock::MockTransport;
+
+    /// Top-level function that creates a project controller.
+    /// Automatically establishes WS connections with endpoints given in `conf`.
+    pub async fn setup(conf:&SetupConfiguration) -> StrongHandle<project::Data> {
+        let file_manager_transport = conf.connect_fm().await;
+        project::new(file_manager_transport)
+    }
+
+    /// Configuration data necessary to setup the project controller.
+    pub struct SetupConfiguration {
+        /// URL of the websocket endpoint of the file manager server.
+        pub file_manager_endpoint:String,
+    }
+
+    impl SetupConfiguration {
+        /// Establishes connection with the remote file manager endpoint.
+        pub async fn connect_fm(&self) -> MockTransport {
+            // TODO [mwu] should not return mock transport but a real class
+            //      implementing the websocket-based transport.
+            todo!()
+        }
+    }
+}
+
+mod project {
+    //! Project controller.
+    use super::*;
+    use json_rpc::Transport;
+
+    /// Create a new project controller.
+    pub fn new(file_manager_transport:impl Transport + 'static) -> StrongHandle<project::Data> {
+        let file_manager = file_manager_client::Client::new(file_manager_transport);
+        let ret = Data {
+            file_manager,
+            modules: default(),
+        };
+        strong(ret)
+    }
+
+    /// Project controller's state.
+    #[derive(Debug)]
+    pub struct Data {
+        file_manager : file_manager_client::Client,
+        modules      : HashMap<module::Location, module::WeakHandle>,
+    }
+
+    impl Data {
+        /// Returns handle to given module's controller if already present.
+        pub fn lookup_module(&mut self, loc:&module::Location) -> Option<module::StrongHandle> {
+            let weak = self.modules.get(loc)?;
+            match weak.upgrade() {
+                Some(strong) => Some(strong),
+                None => {
+                    self.modules.remove(loc);
+                    None
+                }
+            }
+        }
+        /// Stores given module controller handle and returns it.
+        ///
+        /// Note: handle stored in the project controller is weak, so the caller
+        /// remains responsible for managing the module's controller lifetime.
+        pub fn insert_module(&mut self, data:module::Data) -> module::StrongHandle {
+            let path   = data.path.clone();
+            let module = strong(data);
+            // TODO what if already exists?
```

Comment:
Then return the existing module. 

This is why we are playing with Rc handles here.